File: platform/lib-java/src/main/java/com/canary/platform/lib/observability/CanaryMetrics.java
package com.canary.platform.lib.observability;

import io.micrometer.core.instrument.Counter;
import io.micrometer.core.instrument.MeterRegistry;
import io.micrometer.core.instrument.Tags;
import io.micrometer.core.instrument.Timer;

import java.time.Duration;

public class CanaryMetrics {

    private static final String COUNTER_NAME = "canary_request_total";
    private static final String TIMER_NAME = "canary_request_duration_seconds";
    private static final String SHADOW_NAME = "canary_shadow_mismatch_total";

    private final MeterRegistry registry;
    private final String serviceName;

    public CanaryMetrics(MeterRegistry registry, String serviceName) {
        this.registry = registry;
        this.serviceName = serviceName;
    }

    public void recordHttp(String target, String outcome, Duration duration) {
        record("http", target, outcome, duration);
    }

    public void recordKafka(String target, String outcome, Duration duration) {
        record("kafka", target, outcome, duration);
    }

    public void recordRestate(String target, String outcome, Duration duration) {
        record("restate", target, outcome, duration);
    }

    public void recordShadowMismatch(String field) {
        Counter.builder(SHADOW_NAME)
                .tags("service", serviceName, "field", field)
                .register(registry)
                .increment();
    }

    private void record(String substrate, String target, String outcome, Duration duration) {
        String lane = CanaryLaneTag.current();
        Tags counterTags = Tags.of(
                "substrate", substrate,
                "service", serviceName,
                "lane", lane,
                "outcome", outcome,
                "target", target);
        Counter.builder(COUNTER_NAME).tags(counterTags).register(registry).increment();

        Tags timerTags = Tags.of(
                "substrate", substrate,
                "service", serviceName,
                "lane", lane,
                "target", target);
        Timer.builder(TIMER_NAME)
                .tags(timerTags)
                .publishPercentileHistogram()
                .register(registry)
                .record(duration);
    }
}

File: platform/lib-node/src/observability/canary-metrics.ts
import { Counter, Histogram, Registry, register as defaultRegister } from "prom-client";
import { currentLane } from "./canary-lane-tag.js";

export type Substrate = "http" | "kafka" | "restate";
export type Outcome = "success" | "client_error" | "server_error";

const COUNTER_NAME = "canary_request_total";
const HISTOGRAM_NAME = "canary_request_duration_seconds";
const SHADOW_NAME = "canary_shadow_mismatch_total";

export class CanaryMetrics {
  private readonly counter: Counter<string>;
  private readonly histogram: Histogram<string>;
  private readonly shadow: Counter<string>;

  constructor(
    private readonly serviceName: string,
    private readonly registry: Registry = defaultRegister,
  ) {
    this.counter = new Counter({
      name: COUNTER_NAME,
      help: "Total canary-tagged requests per substrate/service/lane.",
      labelNames: ["substrate", "service", "lane", "outcome", "target"],
      registers: [registry],
    });
    this.histogram = new Histogram({
      name: HISTOGRAM_NAME,
      help: "Canary-tagged request duration in seconds.",
      labelNames: ["substrate", "service", "lane", "target"],
      registers: [registry],
    });
    this.shadow = new Counter({
      name: SHADOW_NAME,
      help: "Canary shadow-read field mismatches.",
      labelNames: ["service", "field"],
      registers: [registry],
    });
  }

  recordHttp(target: string, outcome: Outcome, durationSeconds: number): void {
    this.record("http", target, outcome, durationSeconds);
  }

  recordKafka(target: string, outcome: Outcome, durationSeconds: number): void {
    this.record("kafka", target, outcome, durationSeconds);
  }

  recordRestate(target: string, outcome: Outcome, durationSeconds: number): void {
    this.record("restate", target, outcome, durationSeconds);
  }

  recordShadowMismatch(field: string): void {
    this.shadow.inc({ service: this.serviceName, field });
  }

  /** For service wiring: hand the registry to the metrics endpoint. */
  getRegistry(): Registry {
    return this.registry;
  }

  private record(substrate: Substrate, target: string, outcome: Outcome, durationSeconds: number): void {
    const lane = currentLane();
    this.counter.inc({
      substrate,
      service: this.serviceName,
      lane,
      outcome,
      target,
    });
    this.histogram.observe(
      { substrate, service: this.serviceName, lane, target },
      durationSeconds,
    );
  }
}
